File: native/ios/AllStakRNModule.m
// RCTBridgeModule bridging AllStakCrashHandler to JS.
//
// SCAFFOLDED: requires React Native iOS project with CocoaPods autolinking
// to verify end-to-end.

#import <React/RCTBridgeModule.h>
#import "AllStakCrashHandler.h"
#import <UIKit/UIKit.h>
#import <QuartzCore/QuartzCore.h>

@interface AllStakRNModule : NSObject <RCTBridgeModule>
@end

@implementation AllStakRNModule

static CFTimeInterval AllStakProcessStartTime = 0;
static BOOL AllStakFrameMonitorStarted = NO;
static CFTimeInterval AllStakLastFrameTimestamp = 0;
static NSInteger AllStakTotalFrames = 0;
static NSInteger AllStakSlowFrames = 0;
static NSInteger AllStakFrozenFrames = 0;
static double AllStakMaxFrameDelayMs = 0;
static CADisplayLink *AllStakDisplayLink = nil;

+ (void)initialize {
    if (self == [AllStakRNModule class]) {
        AllStakProcessStartTime = CACurrentMediaTime();
    }
}

RCT_EXPORT_MODULE(AllStakNative);

- (void)allstakFrameTick:(CADisplayLink *)displayLink {
    if (AllStakLastFrameTimestamp > 0) {
        double expectedFrameMs = displayLink.duration > 0 ? displayLink.duration * 1000.0 : 16.7;
        double frameDeltaMs = (displayLink.timestamp - AllStakLastFrameTimestamp) * 1000.0;
        double delayMs = MAX(0, frameDeltaMs - expectedFrameMs);
        AllStakTotalFrames += 1;
        if (delayMs > 50.0) AllStakSlowFrames += 1;
        if (delayMs > 700.0) AllStakFrozenFrames += 1;
        if (delayMs > AllStakMaxFrameDelayMs) AllStakMaxFrameDelayMs = delayMs;
    }
    AllStakLastFrameTimestamp = displayLink.timestamp;
}

- (void)ensureFrameMonitorStarted {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (AllStakFrameMonitorStarted) return;
        AllStakFrameMonitorStarted = YES;
        AllStakDisplayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(allstakFrameTick:)];
        [AllStakDisplayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
    });
}

RCT_EXPORT_METHOD(install:(NSString *)release
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [AllStakCrashHandler installWithRelease:release];
    [self ensureFrameMonitorStarted];
    resolve(@YES);
}

RCT_EXPORT_METHOD(getPerformanceSnapshot:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [self ensureFrameMonitorStarted];
    dispatch_async(dispatch_get_main_queue(), ^{
        NSDictionary *snapshot = @{
            @"native_app_start_ms": @((CACurrentMediaTime() - AllStakProcessStartTime) * 1000.0),
            @"total_frames": @(AllStakTotalFrames),
            @"slow_frames": @(AllStakSlowFrames),
            @"frozen_frames": @(AllStakFrozenFrames),
            @"max_frame_delay_ms": @(AllStakMaxFrameDelayMs)
        };
        AllStakTotalFrames = 0;
        AllStakSlowFrames = 0;
        AllStakFrozenFrames = 0;
        AllStakMaxFrameDelayMs = 0;
        resolve(snapshot);
    });
}

RCT_EXPORT_METHOD(drainPendingCrash:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *json = [AllStakCrashHandler drainPendingCrash];
    resolve(json ?: [NSNull null]);
}

RCT_EXPORT_METHOD(captureScreenshot:(NSDictionary *)options
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(dispatch_get_main_queue(), ^{
        @try {
            UIWindow *window = nil;
            if (@available(iOS 13.0, *)) {
                for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
                    if (scene.activationState != UISceneActivationStateForegroundActive ||
                        ![scene isKindOfClass:UIWindowScene.class]) {
                        continue;
                    }
                    UIWindowScene *windowScene = (UIWindowScene *)scene;
                    for (UIWindow *candidate in windowScene.windows) {
                        if (candidate.isKeyWindow) {
                            window = candidate;
                            break;
                        }
                    }
                    if (window != nil) break;
                }
            }
            if (window == nil) {
                window = UIApplication.sharedApplication.keyWindow;
            }
            if (window == nil) {
                reject(@"screenshot-unavailable", @"No active window available", nil);
                return;
            }

            NSString *format = [options[@"format"] isKindOfClass:NSString.class]
                ? [options[@"format"] lowercaseString]
                : @"jpg";
            CGFloat quality = [options[@"quality"] respondsToSelector:@selector(doubleValue)]
                ? [options[@"quality"] doubleValue]
                : 0.7;
            quality = MAX(0.0, MIN(1.0, quality));

            CGSize size = window.bounds.size;
            UIGraphicsImageRendererFormat *rendererFormat = [UIGraphicsImageRendererFormat defaultFormat];
            rendererFormat.scale = window.screen.scale;
            rendererFormat.opaque = NO;
            UIGraphicsImageRenderer *renderer = [[UIGraphicsImageRenderer alloc] initWithSize:size format:rendererFormat];
            UIImage *image = [renderer imageWithActions:^(UIGraphicsImageRendererContext *context) {
                BOOL rendered = [window drawViewHierarchyInRect:window.bounds afterScreenUpdates:NO];
                if (!rendered) {
                    [window.layer renderInContext:context.CGContext];
                }
            }];

            NSData *data = nil;
            NSString *contentType = @"image/jpeg";
            if ([format isEqualToString:@"png"]) {
                data = UIImagePNGRepresentation(image);
                contentType = @"image/png";
            } else {
                data = UIImageJPEGRepresentation(image, quality);
            }
            if (data == nil) {
                reject(@"screenshot-failed", @"Could not encode screenshot", nil);
                return;
            }
            NSNumber *maxBytes = options[@"maxBytes"];
            if ([maxBytes respondsToSelector:@selector(unsignedIntegerValue)] &&
                data.length > [maxBytes unsignedIntegerValue]) {
                reject(@"screenshot-too-large", @"Screenshot exceeds configured maxBytes", nil);
                return;
            }

            resolve(@{
                @"dataBase64": [data base64EncodedStringWithOptions:0],
                @"contentType": contentType,
                @"width": @(lrint(size.width)),
                @"height": @(lrint(size.height)),
                @"sizeBytes": @(data.length)
            });
        } @catch (NSException *exception) {
            reject(@"screenshot-failed", exception.reason ?: @"Screenshot capture failed", nil);
        }
    });
}

// DEV-ONLY: deliberately crash the app to verify the crash-capture flow.
// Throws an NSException synchronously on the next runloop tick. Never
// expose this in user-facing UI of a production app — it kills the
// process. The JS-side wrapper (`AllStak.__devCrashIos__`) is documented
// as DEV_ONLY and is excluded from the public DX surface.
RCT_EXPORT_METHOD(__devTriggerCrash:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(dispatch_get_main_queue(), ^{
        @throw [NSException
            exceptionWithName:@"AllStakDevCrash"
            reason:@"Dev-only: deliberate native crash to verify capture"
            userInfo:nil];
    });
    resolve(@YES);
}

@end

File: native/android/src/main/java/io/allstak/rn/AllStakRNModule.java
package io.allstak.rn;

import androidx.annotation.NonNull;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.os.Build;
import android.os.SystemClock;
import android.util.Base64;
import android.view.Choreographer;
import android.view.View;

import java.io.ByteArrayOutputStream;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.UiThreadUtil;
import com.facebook.react.bridge.WritableMap;

/**
 * React Native module bridging {@link AllStakCrashHandler}.
 *
 *  JS side:
 *     NativeModules.AllStakNative.drainPendingCrash().then(json => ...)
 *
 * SCAFFOLDED — compiles against React Native; requires a bare RN app with
 * autolinking to verify end-to-end.
 */
public class AllStakRNModule extends ReactContextBaseJavaModule {
    private static final long PROCESS_START_MS = SystemClock.uptimeMillis();
    private static final double SLOW_FRAME_MS = 50.0d;
    private static final double FROZEN_FRAME_MS = 700.0d;

    private boolean frameMonitorStarted = false;
    private long lastFrameNanos = 0L;
    private int totalFrames = 0;
    private int slowFrames = 0;
    private int frozenFrames = 0;
    private double maxFrameDelayMs = 0.0d;

    public AllStakRNModule(ReactApplicationContext ctx) { super(ctx); }

    @NonNull
    @Override
    public String getName() { return "AllStakNative"; }

    private synchronized void ensureFrameMonitorStarted() {
        if (frameMonitorStarted) return;
        frameMonitorStarted = true;
        UiThreadUtil.runOnUiThread(() -> {
            Choreographer.getInstance().postFrameCallback(new Choreographer.FrameCallback() {
                @Override
                public void doFrame(long frameTimeNanos) {
                    synchronized (AllStakRNModule.this) {
                        if (lastFrameNanos > 0L) {
                            double frameDeltaMs = (frameTimeNanos - lastFrameNanos) / 1_000_000.0d;
                            double delayMs = Math.max(0.0d, frameDeltaMs - 16.7d);
                            totalFrames += 1;
                            if (delayMs > SLOW_FRAME_MS) slowFrames += 1;
                            if (delayMs > FROZEN_FRAME_MS) frozenFrames += 1;
                            if (delayMs > maxFrameDelayMs) maxFrameDelayMs = delayMs;
                        }
                        lastFrameNanos = frameTimeNanos;
                    }
                    Choreographer.getInstance().postFrameCallback(this);
                }
            });
        });
    }

    @ReactMethod
    public void install(String release, Promise promise) {
        try {
            AllStakCrashHandler.install(getReactApplicationContext(), release);
            ensureFrameMonitorStarted();
            promise.resolve(true);
        } catch (Throwable t) {
            promise.reject("install-failed", t);
        }
    }

    @ReactMethod
    public void getPerformanceSnapshot(Promise promise) {
        try {
            ensureFrameMonitorStarted();
            WritableMap result = Arguments.createMap();
            synchronized (this) {
                result.putDouble("native_app_start_ms", Math.max(0L, SystemClock.uptimeMillis() - PROCESS_START_MS));
                result.putInt("total_frames", totalFrames);
                result.putInt("slow_frames", slowFrames);
                result.putInt("frozen_frames", frozenFrames);
                result.putDouble("max_frame_delay_ms", maxFrameDelayMs);
                totalFrames = 0;
                slowFrames = 0;
                frozenFrames = 0;
                maxFrameDelayMs = 0.0d;
            }
            promise.resolve(result);
        } catch (Throwable t) {
            promise.reject("performance-snapshot-failed", t);
        }
    }

    @ReactMethod
    public void drainPendingCrash(Promise promise) {
        try {
            String json = AllStakCrashHandler.drainPendingCrash(getReactApplicationContext());
            promise.resolve(json);
        } catch (Throwable t) {
            promise.reject("drain-failed", t);
        }
    }

    @ReactMethod
    public void captureScreenshot(ReadableMap options, Promise promise) {
        Activity activity = getCurrentActivity();
        if (activity == null) {
            promise.reject("screenshot-unavailable", "No current Activity available");
            return;
        }

        activity.runOnUiThread(() -> {
            try {
                View root = activity.getWindow() != null
                    ? activity.getWindow().getDecorView().getRootView()
                    : null;
                if (root == null || root.getWidth() <= 0 || root.getHeight() <= 0) {
                    promise.reject("screenshot-unavailable", "No visible root view available");
                    return;
                }

                Bitmap bitmap = Bitmap.createBitmap(root.getWidth(), root.getHeight(), Bitmap.Config.ARGB_8888);
                Canvas canvas = new Canvas(bitmap);
                root.draw(canvas);

                String format = options != null && options.hasKey("format")
                    ? options.getString("format")
                    : "jpg";
                double qualityValue = options != null && options.hasKey("quality")
                    ? options.getDouble("quality")
                    : 0.7d;
                int quality = (int) Math.round(Math.max(0d, Math.min(1d, qualityValue)) * 100d);

                Bitmap.CompressFormat compressFormat = Bitmap.CompressFormat.JPEG;
                String contentType = "image/jpeg";
                if ("png".equalsIgnoreCase(format)) {
                    compressFormat = Bitmap.CompressFormat.PNG;
                    contentType = "image/png";
                    quality = 100;
                } else if ("webp".equalsIgnoreCase(format)) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                        compressFormat = Bitmap.CompressFormat.WEBP_LOSSY;
                    } else {
                        compressFormat = Bitmap.CompressFormat.WEBP;
                    }
                    contentType = "image/webp";
                }

                ByteArrayOutputStream out = new ByteArrayOutputStream();
                boolean encoded = bitmap.compress(compressFormat, quality, out);
                bitmap.recycle();
                if (!encoded) {
                    promise.reject("screenshot-failed", "Could not encode screenshot");
                    return;
                }

                byte[] bytes = out.toByteArray();
                if (options != null && options.hasKey("maxBytes")) {
                    int maxBytes = options.getInt("maxBytes");
                    if (maxBytes > 0 && bytes.length > maxBytes) {
                        promise.reject("screenshot-too-large", "Screenshot exceeds configured maxBytes");
                        return;
                    }
                }

                WritableMap result = Arguments.createMap();
                result.putString("dataBase64", Base64.encodeToString(bytes, Base64.NO_WRAP));
                result.putString("contentType", contentType);
                result.putInt("width", root.getWidth());
                result.putInt("height", root.getHeight());
                result.putInt("sizeBytes", bytes.length);
                promise.resolve(result);
            } catch (Throwable t) {
                promise.reject("screenshot-failed", t);
            }
        });
    }

    /**
     * DEV-ONLY: deliberately crash the app from a non-React thread so the
     * Thread.UncaughtExceptionHandler in {@link AllStakCrashHandler} fires.
     * Never call this from production code — it terminates the process.
     * The JS wrapper {@code AllStak.__devCrashAndroid__} is documented as
     * DEV_ONLY and is excluded from the public DX surface.
     */
    @ReactMethod
    public void __devTriggerCrash(Promise promise) {
        promise.resolve(true);
        new Thread(() -> {
            try { Thread.sleep(50); } catch (InterruptedException ignored) {}
            throw new RuntimeException(
                "AllStakDevCrash: Dev-only deliberate native crash to verify capture");
        }, "AllStakDevCrashThread").start();
    }
}
